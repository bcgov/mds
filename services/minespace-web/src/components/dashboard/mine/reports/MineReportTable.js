/* eslint-disable */
import React from "react";
import { Table } from "antd";
import moment from "moment";
import PropTypes from "prop-types";
import NullScreen from "@/components/common/NullScreen";
import * as Strings from "@/constants/strings";
import { formatDate } from "@/utils/helpers";
import { COLOR } from "@/constants/styles";
import CustomPropTypes from "@/customPropTypes";
import { MineReportActions } from "@/components/dashboard/mine/reports/MineReportActions";

const { errorRed } = COLOR;

/**
 * @class  MinePermitInfo - contains all permit information
 */

const propTypes = {
  mineReports: PropTypes.arrayOf(CustomPropTypes.mineReport).isRequired,
  openEditReportModal: PropTypes.func.isRequired,
  handleEditReport: PropTypes.func.isRequired,
};

const defaultProps = {};

const columns = [
  {
    title: "Report Name",
    dataIndex: "report_name",
    key: "report_name",
    sorter: (a, b) => a.report_name.localeCompare(b.report_name),
    render: (text, record) => (
      <div title="Report Name" style={record.isOverdue ? { color: errorRed } : {}}>
        {record.report_name}
      </div>
    ),
  },
  {
    title: "Compliance Year/Period",
    dataIndex: "submission_year",
    key: "submission_year",
    sorter: (a, b) => (a.submission_year > b.submission_year ? -1 : 1),
    render: (text, record) => (
      <div title="Year" style={record.isOverdue ? { color: errorRed } : {}}>
        {record.submission_year}
      </div>
    ),
  },
  {
    title: "Due",
    dataIndex: "due_date",
    key: "due_date",
    sorter: (a, b) => (moment(a.due_date) > moment(b.due_date) ? -1 : 1),
    render: (text, record) => (
      <div title="Due" style={record.isOverdue ? { color: errorRed } : {}}>
        {formatDate(record.due_date) || Strings.EMPTY_FIELD}
      </div>
    ),
  },
  {
    title: "Received",
    dataIndex: "received_date",
    key: "received_date",
    sorter: (a, b) => (moment(a.received_date) > moment(b.received_date) ? -1 : 1),
    render: (text, record) => (
      <div title="Received" style={record.isOverdue ? { color: errorRed } : {}}>
        {formatDate(record.received_date) || Strings.EMPTY_FIELD}
      </div>
    ),
  },
  {
    title: "",
    dataIndex: "record",
    render: (text, record) => {
      return (
        <div title="" align="right">
          <MineReportActions
            mineReport={record.report}
            openEditReportModal={record.openEditReportModal}
            handleEditReport={record.handleEditReport}
          />
        </div>
      );
    },
  },
];

const transformRowData = (report, openEditReportModal, handleEditReport, handleRemoveReport) => ({
  key: report.report_guid,
  report,
  report_name: report.report_name,
  due_date: report.due_date,
  received_date: report.received_date,
  submission_year: report.submission_year,
  openEditReportModal,
  handleEditReport,
  handleRemoveReport,
});

export const MineReportTable = (props) => (
  <Table
    align="left"
    pagination={false}
    columns={columns}
    locale={{ emptyText: <NullScreen type="reports" /> }}
    dataSource={props.mineReports.map((r) =>
      transformRowData(
        r,
        props.openEditReportModal,
        props.handleEditReport,
        props.handleRemoveReport
      )
    )}
  />
);

MineReportTable.propTypes = propTypes;
MineReportTable.defaultProps = defaultProps;

export default MineReportTable;
