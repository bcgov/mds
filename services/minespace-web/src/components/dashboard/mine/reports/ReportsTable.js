/* eslint-disable */

import React from "react";
import { Table, Button } from "antd";
import moment from "moment";
import PropTypes from "prop-types";
import * as Strings from "@/constants/strings";
import { formatDate } from "@/utils/helpers";
import { EDIT_PENCIL } from "@/constants/assets";
import CustomPropTypes from "@/customPropTypes";
import LinkButton from "@/components/common/LinkButton";
import { truncateFilename } from "@common/utils/helpers";
import { downloadFileFromDocumentManager } from "@common/utils/actionlessNetworkCalls";

const propTypes = {
  mineReports: PropTypes.arrayOf(CustomPropTypes.mineReport).isRequired,
  openEditReportModal: PropTypes.func.isRequired,
  handleEditReport: PropTypes.func.isRequired,
  isLoaded: PropTypes.bool.isRequired,
};

const defaultProps = {};

const columns = [
  {
    title: "Report Name",
    dataIndex: "report_name",
    key: "report_name",
    sorter: (a, b) => a.report_name.localeCompare(b.report_name),
    render: (text, record) => (
      <div title="Report Name" className={record.isOverdue ? "color-error" : ""}>
        {record.report_name}
      </div>
    ),
  },
  {
    title: "Compliance Period",
    dataIndex: "submission_year",
    key: "submission_year",
    sorter: (a, b) => (a.submission_year > b.submission_year ? -1 : 1),
    render: (text, record) => (
      <div title="Compliance Period" className={record.isOverdue ? "color-error" : ""}>
        {record.submission_year}
      </div>
    ),
  },
  {
    title: "Due",
    dataIndex: "due_date",
    key: "due_date",
    //FIXME sorter: (a, b) => (moment(a.due_date) > moment(b.due_date) ? -1 : 1),
    render: (due_date, record) => (
      <div title="Due" className={record.isOverdue ? "color-error" : ""}>
        {formatDate(due_date) || Strings.EMPTY_FIELD}
      </div>
    ),
  },
  {
    title: "Submitted On",
    dataIndex: "received_date",
    key: "received_date",
    //FIXME sorter: (a, b) =>
    //  moment(a.received_date) > moment(b.received_date) ? -1 : 1,
    render: (received_date, record) => (
      <div title="Submitted On" className={record.isOverdue ? "color-error" : ""}>
        {formatDate(received_date) || Strings.EMPTY_FIELD}
      </div>
    ),
  },
  {
    title: "Requested By",
    dataIndex: "created_by_idir",
    key: "created_by_idir",
    sorter: (a, b) => a.created_by_idir.localeCompare(b.created_by_idir),
    render: (text, record) => (
      <div title="Requested By" className={record.isOverdue ? "color-error" : ""}>
        {text || Strings.EMPTY_FIELD}
      </div>
    ),
  },
  {
    title: "Documents",
    dataIndex: "mine_report_submissions",
    key: "mine_report_submissions",
    render: (text, record) => {
      return (
        <div
          title="Documents"
          className={record.isOverdue ? "color-error cap-col-height" : "cap-col-height"}
        >
          {text.map((sub) =>
            sub.documents.map((doc) => (
              <div>
                -{" "}
                <LinkButton
                  key={doc.mine_document_guid}
                  onClick={() => downloadFileFromDocumentManager(doc)}
                  title={doc.document_name}
                >
                  {truncateFilename(doc.document_name)}
                </LinkButton>
              </div>
            ))
          ) || Strings.EMPTY_FIELD}
        </div>
      );
    },
  },
  {
    title: "",
    dataIndex: "report",
    render: (text, record) => {
      return (
        <div title="" align="right">
          <Button type="link" onClick={(event) => record.openEditReportModal(event, text)}>
            <img src={EDIT_PENCIL} alt="Edit" />
          </Button>
        </div>
      );
    },
  },
];

const transformRowData = (report, openEditReportModal, handleEditReport, handleRemoveReport) => ({
  key: report.report_guid,
  report,
  ...report,
  openEditReportModal,
  handleEditReport,
  handleRemoveReport,
});

export const ReportsTable = (props) => {
  return (
    <Table
      size="small"
      pagination={false}
      loading={!props.isLoaded}
      columns={columns}
      locale={{ emptyText: "This mine has no report data." }}
      dataSource={props.mineReports.map((record) =>
        transformRowData(
          record,
          props.openEditReportModal,
          props.handleEditReport,
          props.handleRemoveReport
        )
      )}
    />
  );
};

ReportsTable.propTypes = propTypes;
ReportsTable.defaultProps = defaultProps;

export default ReportsTable;
