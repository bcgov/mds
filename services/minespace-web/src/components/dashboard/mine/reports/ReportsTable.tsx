import React, { FC } from "react";
import { connect } from "react-redux";
import { Badge, Button, TablePaginationConfig } from "antd";
import {
  formatComplianceCodeValueOrLabel,
  truncateFilename,
} from "@mds/common/redux/utils/helpers";
import { downloadFileFromDocumentManager } from "@common/utils/actionlessNetworkCalls";
import { getMineReportDefinitionHash } from "@mds/common/redux/selectors/staticContentSelectors";
import * as Strings from "@/constants/strings";
import { EDIT_PENCIL } from "@/constants/assets";
import LinkButton from "@/components/common/LinkButton";
import AuthorizationWrapper from "@/components/common/wrappers/AuthorizationWrapper";
import {
  Feature,
  IMineReport,
  MINE_REPORT_STATUS_HASH,
  MINE_REPORT_SUBMISSION_CODES,
} from "@mds/common";
import { useFeatureFlag } from "@mds/common/providers/featureFlags/useFeatureFlag";
import {
  renderActionsColumn,
  renderTextColumn,
} from "@mds/common/components/common/CoreTableCommonColumns";
import EyeOutlined from "@ant-design/icons/EyeOutlined";
import { ColumnsType } from "antd/es/table";
import CoreTable from "@mds/common/components/common/CoreTable";

interface ReportsTableProps {
  mineReports: IMineReport[];
  mineReportDefinitionHash: any;
  openEditReportModal: (event: React.MouseEvent, record: IMineReport) => void;
  openReport: (record: IMineReport) => void;
  isLoaded: boolean;
}

const DEFAULT_PAGE_SIZE = 10;

export const reportStatusSeverity = (status: MINE_REPORT_SUBMISSION_CODES) => {
  switch (status) {
    case MINE_REPORT_SUBMISSION_CODES.REQ:
      return "processing";
    case MINE_REPORT_SUBMISSION_CODES.ACC:
      return "success";
    case MINE_REPORT_SUBMISSION_CODES.REC:
      return "warning";
    case MINE_REPORT_SUBMISSION_CODES.NRQ:
      return "default";
    case MINE_REPORT_SUBMISSION_CODES.INI:
      return "default";
    default:
      return "default";
  }
};

export const ReportsTable: FC<ReportsTableProps> = (props) => {
  const { isFeatureEnabled } = useFeatureFlag();

  const actions = [
    {
      key: "view",
      label: "View",
      clickFunction: (_event, record) => {
        props.openReport(record);
      },
      icon: <EyeOutlined />,
    },
  ];

  let columns: ColumnsType<IMineReport> = [
    renderTextColumn("report_name", "Report Name", true),
    {
      title: "Code Section",
      key: "code_section",
      render: (record) => (
        <div title="Code Section">
          {formatComplianceCodeValueOrLabel(
            props.mineReportDefinitionHash[record.mine_report_definition_guid]
              .compliance_articles[0],
            false
          )}
        </div>
      ),
    },
    renderTextColumn("submission_year", "Compliance Year", true, null, 5),
    renderTextColumn("due_date", "Due", true, null, 5),
    renderTextColumn("received_date", "Submitted On", true),
    renderTextColumn("created_by_idir", "Requested By", true),
    {
      title: "Documents",
      dataIndex: "mine_report_submissions",
      render: (text) => (
        <div title="Documents" className="cap-col-height">
          {(text &&
            text.length > 0 &&
            text[text.length - 1].documents &&
            text[text.length - 1].documents.length > 0 &&
            text[text.length - 1].documents.map((document) => (
              <LinkButton
                key={document.mine_document_guid}
                onClick={() => downloadFileFromDocumentManager(document)}
                title={document.document_name}
              >
                {truncateFilename(document.document_name)}
                <br />
              </LinkButton>
            ))) ||
            Strings.EMPTY_FIELD}
        </div>
      ),
    },
    {
      title: "",
      dataIndex: "edit",
      render: (text, record) => {
        return (
          <div>
            <AuthorizationWrapper>
              <Button type="link" onClick={(event) => props.openEditReportModal(event, record)}>
                <img src={EDIT_PENCIL} alt="Edit" />
              </Button>
            </AuthorizationWrapper>
          </div>
        );
      },
    },
  ];

  if (isFeatureEnabled(Feature.CODE_REQUIRED_REPORTS)) {
    columns = columns.filter((column) => !["", "Documents"].includes(column.title as string));
    const statusColumn = {
      title: "Status",
      dataIndex: "status",
      sorter: (a, b) => a.status.localeCompare(b.status),
      render: (text: MINE_REPORT_SUBMISSION_CODES) => {
        return <Badge status={reportStatusSeverity(text)} text={MINE_REPORT_STATUS_HASH[text]} />;
      },
    };

    const newColumns = [statusColumn, renderActionsColumn({ actions })];
    columns = [...columns, ...newColumns];
  }

  const transformRowData = (reports: IMineReport[]): IMineReport[] =>
    reports.map((report) => {
      const { mine_report_submissions } = report;
      const latestSubmission = mine_report_submissions?.[mine_report_submissions?.length - 1];

      return {
        ...report,
        status: latestSubmission?.mine_report_submission_status_code ?? "",
      };
    });

  const pagination: TablePaginationConfig = {
    defaultPageSize: DEFAULT_PAGE_SIZE,
    total: props.mineReports.length,
    position: ["bottomCenter"],
  };

  return (
    <CoreTable
      size={"small"}
      loading={!props.isLoaded}
      columns={columns}
      rowKey={(record) => record.mine_report_guid}
      emptyText="This mine has no report data."
      dataSource={transformRowData(props.mineReports)}
      pagination={pagination}
    />
  );
};

const mapStateToProps = (state) => ({
  mineReportDefinitionHash: getMineReportDefinitionHash(state),
});

export default connect(mapStateToProps, null)(ReportsTable);
