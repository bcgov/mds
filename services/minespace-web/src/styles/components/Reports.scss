.submit-report-button {
  position: absolute;
  top: 10px;
  right: 0;
}

.report-title {
  margin-bottom: 36px !important;
  margin-top: 16px;
}

.report-steps {
  margin-bottom: 36px;
  margin-top: 36px;
}

.report-link {
  text-decoration: none;
  line-height: 32px;
}

.report-instructions {
  color: rgba(49, 49, 50, 0.55);
}

.report-info-box {
  height: 100%;
  padding: 16px;
  border: 1px solid $background;
}
