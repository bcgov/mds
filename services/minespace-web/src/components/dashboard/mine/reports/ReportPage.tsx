import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link, useParams } from "react-router-dom";
import { getFormSubmitErrors, getFormValues, isDirty, submit } from "redux-form";

import { Button, Col, Row, Typography } from "antd";
import ArrowLeftOutlined from "@ant-design/icons/ArrowLeftOutlined";

import { FORM } from "@mds/common/constants/forms";
import ReportDetailsForm from "@mds/common/components/reports/ReportDetailsForm";
import Loading from "@/components/common/Loading";
import { getMineById } from "@mds/common/redux/selectors/mineSelectors";
import { fetchMineRecordById } from "@mds/common/redux/actionCreators/mineActionCreator";
import {
  IMine,
  IMineReportSubmission,
  MINE_REPORT_STATUS_HASH,
  MINE_REPORT_SUBMISSION_CODES,
} from "@mds/common";
import Callout from "@mds/common/components/common/Callout";
import { reportStatusSeverity } from "./ReportsTable";
import {
  createReportSubmission,
  fetchLatestReportSubmission,
  getLatestReportSubmission,
} from "@mds/common/components/reports/reportSubmissionSlice";

const ReportPage = () => {
  const dispatch = useDispatch();
  const { mineGuid, reportGuid } = useParams<{ mineGuid: string; reportGuid: string }>();
  const latestSubmission: IMineReportSubmission = useSelector((state) =>
    getLatestReportSubmission(state, reportGuid)
  );
  const mine: IMine = useSelector((state) => getMineById(state, mineGuid));
  const [loaded, setIsLoaded] = useState(Boolean(latestSubmission && mine));
  const [isEditMode, setIsEditMode] = useState(false);

  // get form data so we can submit it outside of the form
  const formErrors = useSelector(getFormSubmitErrors(FORM.VIEW_EDIT_REPORT));
  const formValues = useSelector(getFormValues(FORM.VIEW_EDIT_REPORT));
  const isFormDirty = useSelector(isDirty(FORM.VIEW_EDIT_REPORT));

  useEffect(() => {
    let isMounted = true;
    const isLoaded = Boolean(mine && latestSubmission);
    if (isMounted) {
      setIsLoaded(isLoaded);
    }
    return () => (isMounted = false);
  }, [mine, latestSubmission]);

  useEffect(() => {
    if (!mine || mine.mine_guid !== mineGuid) {
      dispatch(fetchMineRecordById(mineGuid));
    }

    if (!latestSubmission || latestSubmission.mine_report_guid !== reportGuid) {
      dispatch(fetchLatestReportSubmission({ mine_report_guid: reportGuid }));
    }
  }, [reportGuid, mineGuid]);

  const handleUpdateReport = (values) => {
    dispatch(createReportSubmission({ ...values, mine_guid: mineGuid })).then((response) => {
      if (response.payload) {
        setIsEditMode(false);
      }
    });
  };

  const handleOtherSubmitButton = () => {
    dispatch(submit(FORM.VIEW_EDIT_REPORT));
    if (!formErrors) {
      handleUpdateReport(formValues);
    }
  };

  const status = MINE_REPORT_STATUS_HASH[latestSubmission?.mine_report_submission_status_code];

  return (
    (loaded && (
      <div>
        <Row>
          <Col span={24}>
            <Typography.Title>Report - {mine.mine_name}</Typography.Title>
          </Col>
        </Row>
        <Row>
          <Col span={24}>
            <Link to={GLOBAL_ROUTES?.MINE_DASHBOARD.dynamicRoute(mineGuid, "reports")}>
              <ArrowLeftOutlined className="padding-sm--right" />
              {`Back to: Reports page`}
            </Link>
          </Col>
        </Row>
        <Row align="middle" justify="space-between">
          <Typography.Title level={2}>{latestSubmission.report_name}</Typography.Title>
          {isEditMode && (
            <Button type="primary" disabled={!isFormDirty} onClick={handleOtherSubmitButton}>
              Save Changes
            </Button>
          )}
        </Row>

        {!isEditMode && status && (
          <Callout
            title={`Submission ${status}`}
            message={
              <>
                <p>Your report has been {status}.</p>
                <p>
                  {latestSubmission.submission_date} by {latestSubmission.submitter_name}
                </p>
              </>
            }
            severity={reportStatusSeverity(
              latestSubmission.mine_report_submission_status_code as MINE_REPORT_SUBMISSION_CODES
            )}
          />
        )}
        <ReportDetailsForm
          mineGuid={mineGuid}
          initialValues={latestSubmission}
          handleSubmit={handleUpdateReport}
          isEditMode={isEditMode}
          formButtons={
            isEditMode ? (
              <Button htmlType="submit" disabled={!isFormDirty} type="primary">
                Save Changes
              </Button>
            ) : null
          }
        />
      </div>
    )) || <Loading />
  );
};

export default ReportPage;
