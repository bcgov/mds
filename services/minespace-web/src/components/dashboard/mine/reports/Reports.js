import React, { Component } from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { Row, Col, Typography, Button, Icon } from "antd";
import moment from "moment";
import PropTypes from "prop-types";
import {
  createMineReport,
  fetchMineReports,
  updateMineReport,
} from "@common/actionCreators/reportActionCreator";
import { openModal, closeModal } from "@common/actions/modalActions";
import { getMineReports } from "@common/selectors/reportSelectors";
import { getMineReportDefinitionOptions } from "@common/reducers/staticContentReducer";
import CustomPropTypes from "@/customPropTypes";
import ReportsTable from "@/components/dashboard/mine/reports/ReportsTable";
import TableSummaryCard from "@/components/common/TableSummaryCard";
import { modalConfig } from "@/components/modalContent/config";
import AuthorizationWrapper from "@/components/common/wrappers/AuthorizationWrapper";

const { Paragraph, Title, Text } = Typography;

const propTypes = {
  mine: CustomPropTypes.mine.isRequired,
  mineReports: PropTypes.arrayOf(CustomPropTypes.mineReport).isRequired,
  // This IS being used.
  // eslint-disable-next-line
  mineReportDefinitionOptions: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.any)).isRequired,
  updateMineReport: PropTypes.func.isRequired,
  createMineReport: PropTypes.func.isRequired,
  fetchMineReports: PropTypes.func.isRequired,
  openModal: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
};

export class Reports extends Component {
  state = { isLoaded: false, selectedMineReportGuid: null, reportsDue: 0, reportsSubmitted: 0 };

  componentDidMount() {
    this.props.fetchMineReports(this.props.mine.mine_guid).then(() => {
      this.setState({ isLoaded: true });
    });
  }

  componentWillReceiveProps(nextProps) {
    const reportsSubmitted = nextProps.mineReports.filter(
      (report) => report.mine_report_submissions.length > 0
    ).length;
    const reportsDue = nextProps.mineReports.filter(
      (report) => report.mine_report_submissions.length === 0 && report.due_date
    ).length;

    this.setState({ reportsSubmitted, reportsDue });
  }

  handleAddReport = (values) => {
    const formValues = values;
    if (values.mine_report_submissions !== undefined) {
      formValues.received_date = moment().format("YYYY-MM-DD");
    }
    this.props
      .createMineReport(this.props.mine.mine_guid, formValues)
      .then(() => this.props.closeModal())
      .then(() => this.props.fetchMineReports(this.props.mine.mine_guid));
  };

  handleEditReport = (values) => {
    this.props
      .updateMineReport(this.props.mine.mine_guid, this.state.selectedMineReportGuid, values)
      .then(() => this.props.closeModal())
      .then(() => this.props.fetchMineReports(this.props.mine.mine_guid));
  };

  openAddReportModal = (event) => {
    event.preventDefault();
    this.props.openModal({
      props: {
        onSubmit: this.handleAddReport,
        title: "Add Report",
        mineGuid: this.props.mine.mine_guid,
        width: "40vw",
      },
      content: modalConfig.ADD_REPORT,
    });
  };

  openEditReportModal = (event, report) => {
    this.setState({ selectedMineReportGuid: report.mine_report_guid });
    event.preventDefault();
    this.props.openModal({
      props: {
        initialValues: report,
        onSubmit: this.handleEditReport,
        title: `Add Documents to: ${report.report_name}`,
        mineGuid: this.props.mine.mine_guid,
        width: "40vw",
      },
      content: modalConfig.EDIT_REPORT,
    });
  };

  render() {
    return (
      <Row>
        <Col>
          <Row>
            <Col>
              <Title level={4}>Reports</Title>
              <Paragraph>
                This table shows&nbsp;
                <Text className="color-primary" strong>
                  reports
                </Text>
                &nbsp;that have been submitted to the Ministry. If a report is listed but there are
                no files attached, it means the report has not been submitted. This table may not
                show all reports that your mine is required to submit to the Ministry. If you have
                any questions, please check with an EMPR representative.
              </Paragraph>
              <br />
            </Col>
          </Row>
          {this.props.mineReports && this.props.mineReports.length > 0 && (
            <Row type="flex" justify="space-around" gutter={[16, 32]}>
              <Col md={24} lg={8}>
                <TableSummaryCard
                  title="Reports Submitted"
                  content={this.state.reportsSubmitted}
                  icon="check-circle"
                  type="success"
                />
              </Col>
              <Col md={24} lg={8}>
                <TableSummaryCard
                  title="Reports Due"
                  content={this.state.reportsDue}
                  icon="clock-circle"
                  type="error"
                />
              </Col>
            </Row>
          )}
          <Row gutter={[16, 32]}>
            <Col>
              <AuthorizationWrapper>
                <Button
                  style={{ float: "right" }}
                  type="primary"
                  onClick={(event) => this.openAddReportModal(event, this.props.mine.mine_name)}
                >
                  <Icon type="plus-circle" theme="filled" />
                  Submit Report
                </Button>
              </AuthorizationWrapper>
            </Col>
          </Row>
          <Row gutter={[16, 32]}>
            <Col>
              <ReportsTable
                openEditReportModal={this.openEditReportModal}
                mineReports={this.props.mineReports}
                isLoaded={this.state.isLoaded}
              />
            </Col>
          </Row>
        </Col>
      </Row>
    );
  }
}

const mapStateToProps = (state) => ({
  mineReports: getMineReports(state),
  mineReportDefinitionOptions: getMineReportDefinitionOptions(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchMineReports,
      createMineReport,
      updateMineReport,
      openModal,
      closeModal,
    },
    dispatch
  );

Reports.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(Reports);
