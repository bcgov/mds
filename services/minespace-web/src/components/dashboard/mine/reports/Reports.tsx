import React, { FC, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import PlusCircleFilled from "@ant-design/icons/PlusCircleFilled";
import { Button, Col, Row, Typography } from "antd";
import moment from "moment";
import {
  createMineReport,
  fetchMineReports,
  updateMineReport,
} from "@mds/common/redux/actionCreators/reportActionCreator";
import { closeModal, openModal } from "@mds/common/redux/actions/modalActions";
import { getMineReports } from "@mds/common/redux/selectors/reportSelectors";
import ReportsTable from "@/components/dashboard/mine/reports/ReportsTable";
import { modalConfig } from "@/components/modalContent/config";
import AuthorizationWrapper from "@/components/common/wrappers/AuthorizationWrapper";
import { IMine, IMineReport, Feature } from "@mds/common";
import { Link, useHistory } from "react-router-dom";
import * as routes from "@/constants/routes";
import { useFeatureFlag } from "@mds/common/providers/featureFlags/useFeatureFlag";

interface ReportsProps {
  mine: IMine;
}

export const Reports: FC<ReportsProps> = ({ mine, ...props }) => {
  const dispatch = useDispatch();
  const history = useHistory();
  const { isFeatureEnabled } = useFeatureFlag();

  const mineReports: IMineReport[] = useSelector(getMineReports);

  const [isLoaded, setIsLoaded] = useState(false);
  const [report, setReport] = useState(null);

  useEffect(() => {
    let isMounted = true;

    setIsLoaded(false);

    dispatch(fetchMineReports(mine.mine_guid)).then(() => {
      if (isMounted) {
        setIsLoaded(true);
      }
    });

    return () => {
      isMounted = false;
    };
  }, []);

  const handleAddReport = async (values) => {
    const formValues = values;
    if (values.mine_report_submissions && values.mine_report_submissions.length > 0) {
      formValues.received_date = moment().format("YYYY-MM-DD");
    }

    await dispatch(createMineReport(mine.mine_guid, formValues));
    await dispatch(closeModal());
    return dispatch(fetchMineReports(mine.mine_guid));
  };

  const handleEditReport = async (values) => {
    if (!values.mine_report_submissions || values.mine_report_submissions.length === 0) {
      dispatch(closeModal());
      return;
    }

    let payload: any = {
      mine_report_submissions: [
        ...values.mine_report_submissions,
        {
          documents:
            values.mine_report_submissions[values.mine_report_submissions.length - 1].documents,
        },
      ],
    };

    if (
      !report.received_date &&
      values.mine_report_submissions &&
      values.mine_report_submissions.length > 0
    ) {
      payload = { ...payload, received_date: moment().format("YYYY-MM-DD") };
    }
    await dispatch(updateMineReport(mine.mine_guid, report.mine_report_guid, payload));
    await dispatch(closeModal());
    return dispatch(fetchMineReports(mine.mine_guid));
  };

  const openReport = (reportRecord: IMineReport) => {
    history.push(
      routes.REPORT_VIEW_EDIT.dynamicRoute(mine.mine_guid, reportRecord.mine_report_guid)
    );
  };

  const openAddReportModal = (event) => {
    event.preventDefault();
    dispatch(
      openModal({
        props: {
          onSubmit: handleAddReport,
          title: "Add Report",
          mineGuid: mine.mine_guid,
          width: "40vw",
        },
        content: modalConfig.ADD_REPORT,
      })
    );
  };

  const openEditReportModal = (event, report) => {
    event.preventDefault();
    setReport(report);
    dispatch(
      openModal({
        props: {
          onSubmit: handleEditReport,
          title: `Edit Report: ${report.report_name}`,
          mineGuid: mine.mine_guid,
          width: "40vw",
          mineReport: report,
        },
        content: modalConfig.EDIT_REPORT,
      })
    );
  };

  return (
    <Row>
      <Col span={24}>
        <Row>
          <Col span={24}>
            <div className="submit-report-button">
              <AuthorizationWrapper>
                {isFeatureEnabled(Feature.CODE_REQUIRED_REPORTS) ? (
                  <Link to={routes.REPORTS_CREATE_NEW.dynamicRoute(mine.mine_guid)}>
                    <Button style={{ zIndex: 1 }} className="submit-report-button" type="primary">
                      <PlusCircleFilled />
                      Submit Report
                    </Button>
                  </Link>
                ) : (
                  <Button
                    style={{ zIndex: 1 }}
                    className="submit-report-button"
                    type="primary"
                    onClick={(event) => openAddReportModal(event)}
                  >
                    <PlusCircleFilled />
                    Submit Report
                  </Button>
                )}
              </AuthorizationWrapper>
            </div>
            <Typography.Title level={4} className="report-title">
              Reports
            </Typography.Title>
            <Typography.Paragraph>
              This table shows reports from the Health, Safety and Reclamation code that your mine
              has submitted to the Ministry. It also shows reports the Ministry has requested from
              your mine. If you do not see an HSRC report that your mine must submit, click Submit
              Report, choose the report you need to send and then attach the file or files.
            </Typography.Paragraph>
            <Typography.Paragraph>
              Note: Do not use this page to submit reports specified in your permit. Continue to
              email these reports to the Ministry.
            </Typography.Paragraph>
          </Col>
        </Row>
        <Row gutter={[16, 32]}>
          <Col span={24}>
            <ReportsTable
              openReport={openReport}
              openEditReportModal={openEditReportModal}
              mineReports={mineReports}
              isLoaded={isLoaded}
            />
          </Col>
        </Row>
      </Col>
    </Row>
  );
};

export default Reports;
