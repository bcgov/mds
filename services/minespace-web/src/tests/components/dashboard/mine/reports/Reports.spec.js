import React from "react";
import { shallow } from "enzyme";
import { Reports } from "@/components/dashboard/mine/reports/Reports";
import * as MOCK from "@/tests/mocks/dataMocks";

const props = {};
const dispatchProps = {};

const setupProps = () => {
  props.mine = MOCK.MINES.mines[MOCK.MINES.mineIds[0]];
};

const setupDispatchProps = () => {
  dispatchProps.fetchMineReports = jest.fn(() => Promise.resolve());
  dispatchProps.updateMineReport = jest.fn();
  dispatchProps.fetchMineReportDefinitionOptions = jest.fn();
  dispatchProps.openModal = jest.fn();
  dispatchProps.closeModal = jest.fn();
};

beforeEach(() => {
  setupProps();
  setupDispatchProps();
});

describe("Reports", () => {
  it("renders properly", () => {
    const component = shallow(<Reports {...props} {...dispatchProps} />);
    expect(component).toMatchSnapshot();
  });
});
