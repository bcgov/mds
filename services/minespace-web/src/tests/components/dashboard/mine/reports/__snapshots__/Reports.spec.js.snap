// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reports renders properly 1`] = `
<Row
  gutter={0}
>
  <Col>
    <Row
      gutter={0}
    >
      <Col>
        <Title
          level={4}
        >
          Reports
        </Title>
        <Paragraph>
          This table shows 
          <Text
            className="color-primary"
            strong={true}
          >
            reports
          </Text>
           that have been submitted to the Ministry. If a report is listed but there are no files attached, it means the report has not been submitted. This table may not show all reports that your mine is required to submit to the Ministry. If you have any questions, please check with an EMPR representative.
        </Paragraph>
        <br />
      </Col>
    </Row>
    <Row
      gutter={
        Array [
          16,
          32,
        ]
      }
    >
      <Col>
        <Connect(AuthorizationWrapper)>
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            style={
              Object {
                "float": "right",
              }
            }
            type="primary"
          >
            <Icon
              theme="filled"
              type="plus-circle"
            />
            Submit Report
          </Button>
        </Connect(AuthorizationWrapper)>
      </Col>
    </Row>
    <Row
      gutter={
        Array [
          16,
          32,
        ]
      }
    >
      <Col>
        <ReportsTable
          isLoaded={false}
          openEditReportModal={[Function]}
        />
      </Col>
    </Row>
  </Col>
</Row>
`;
